code/modules/antagonists/changeling/evolution_menu.dm

```diff
@@ -0,0 +1,102 @@
+	/// A list containing the typepaths of bought changeling abilities. For use with the UI.
+	var/list/purchased_abilities = list()
+	/// A list containing every purchasable changeling power. Includes its name, description, helptext and cost.
+	var/static/list/ability_list = list()
+
+/datum/action/changeling/evolution_menu/Grant(mob/M)
+	..()
+	if(length(ability_list))
+		return // List is already populated.
+
+	for(var/power_path in cling.purchaseable_powers)
+		var/datum/action/changeling/C = power_path
+		ability_list += list(list(
+			"name" = initial(C.name),
+			"description" = initial(C.desc),
+			"helptext" = initial(C.helptext),
+			"cost" = initial(C.dna_cost),
+			"power_path" = power_path
+		))
+
+/datum/action/changeling/evolution_menu/try_to_sting(mob/user, mob/target)
+	ui_interact(user)
+
+/datum/action/changeling/evolution_menu/ui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = FALSE, datum/tgui/master_ui = null, datum/ui_state/state = GLOB.always_state)
+	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "EvolutionMenu", "Evolution Menu", 480, 574, master_ui, state)
+		ui.set_autoupdate(FALSE)
+		ui.open()
+
+/datum/action/changeling/evolution_menu/ui_data(mob/user)
+	var/list/data = list(
+		"can_respec" = cling.can_respec,
+		"evo_points" = cling.genetic_points,
+		"purchsed_abilities" = purchased_abilities,
+		"view_mode" = view_mode
+	)
+	return data
+
+/datum/action/changeling/evolution_menu/ui_static_data(mob/user)
+	var/list/data = list(
+		"ability_list" = ability_list
+	)
+	return data
+
+/datum/action/changeling/evolution_menu/ui_act(action, list/params)
+	if(..())
+		return
+
+	switch(action)
+		if("readapt")
+			if(!cling.try_respec())
+				return FALSE
+			purchased_abilities.Cut()
+			return TRUE
+
+		if("purchase")
+			var/power_path = text2path(params["power_path"])
+			if(!ispath(power_path) || !try_purchase_power(power_path))
+				return FALSE
+			purchased_abilities |= power_path
+			return TRUE
+
+		if("set_view_mode")
+			var/new_view_mode = text2num(params["mode"])
+			if(!(new_view_mode in list(COMPACT_MODE, EXPANDED_MODE)))
+				return FALSE
+			view_mode = new_view_mode
+			return TRUE
+
+/datum/action/changeling/evolution_menu/proc/try_purchase_power(power_type)
+	if(power_type in purchased_abilities)
+		to_chat(owner, "<span class='warning'>We have already evolved this ability!</span>")
+		return FALSE
+	var/datum/action/changeling/power = power_type
+	if(cling.absorbed_count < initial(power.req_dna))
+		to_chat(owner, "<span class='warning'>We must absorb more victims before we can evolve this ability!</span>")
+		return FALSE
+	if(initial(power.dna_cost) < 0)
+		to_chat(owner, "<span class='warning'>We cannot evolve this ability.</span>")
+		return FALSE
+	if(cling.genetic_points < initial(power.dna_cost))
+		to_chat(owner, "<span class='warning'>We have reached our capacity for abilities.</span>")
```

Comment:
this text might not be true, if an ability cost is 2 and the changeling only has 1 point left.

Reply:
I'm not sure what you mean. In the case that they have 1 point left and the ability costs 2, that's a situation where they won't be able to buy it. 

As a side not, I'm going to change the error message a bit to reflect this, "reached our capacity" is not really correct.